sympy/core/function.py

```diff
@@ -1200,9 +1177,20 @@ def __new__(cls, expr, *variables, **kwargs):
         # expression at all. Note, this cannot check non-symbols like
         # functions and Derivatives as those can be created by intermediate
         # derivatives.
-        if evaluate and all(isinstance(sc[0], Symbol) for sc in variable_count):
```

Comment:
I'm not sure about this change. The motivation in the past is that for non-Symbol variables, we can't automatically assume that certain derivatives are 0. For example A[i].diff(A[j]) is KroneckerDelta(i, j), where `A = IndexedBase('A')`. Similar things happen with matrices. Granted, those things should all be tested, so is this still handling those cases properly?

Reply:
in this PR:
```
>>> A[i].diff(A[j])
KroneckerDelta(i, j)
```
This isn't 0 because these objects share a free symbol:
```
>>> A[i].free_symbols & A[j].free_symbols
{A}
```